File: packages/infrastructure/src/integrations/ai/idea-draft-schema.ts
import type { IdeaAssistContext } from '@collega/application/ai'
import { businessImpactIdsOf, ideaTypeIdsOf } from '@collega/application/ai'
import { Priority } from '@collega/domain/enums'
import { DESCRIPTION_MAX_LENGTH, TITLE_MAX_LENGTH } from '@collega/domain/ideas'

const PRIORITY_NAMES: readonly string[] = Object.values(Priority)

/**
 * Builds the response JSON Schema PER REQUEST from the retrieval result (SPEC/20-feature-ai-idea-
 * assist.md rules 15-18), ported from .NET's `IdeaDraftSchema`.
 *
 * This is the containment mechanism. Because `ideaTypeId` and `businessImpactId` are closed
 * enums of this organization's real, active option ids, an invalid or cross-org classification is
 * structurally impossible - not prompt-discouraged. `additionalProperties` is false and every
 * enum is closed, so `nextQuestion` is the only string the model can author.
 */
export function buildIdeaDraftResponseSchema(context: IdeaAssistContext): Record<string, unknown> {
  return {
    type: 'object',
    properties: {
      inScope: {
        type: 'boolean',
        description:
          "True when the user's latest message could plausibly become an idea of one of this " +
          "organization's active idea types, within any scope statement given. False otherwise.",
      },
      nextQuestion: {
        type: 'string',
        description:
          'One short follow-up question that moves the draft forward. This is the only free ' +
          'text you may write.',
      },
      title: nullable({ type: 'string', maxLength: TITLE_MAX_LENGTH }),
      description: nullable({ type: 'string', maxLength: DESCRIPTION_MAX_LENGTH }),
      ideaTypeId: nullableEnum(
        ideaTypeIdsOf(context),
        "The idea type, chosen from this organization's active types.",
      ),
      businessImpactId: nullableEnum(
        businessImpactIdsOf(context),
        "The business impact, chosen from this organization's active options.",
      ),
      priority: {
        anyOf: [{ type: 'string', enum: PRIORITY_NAMES }, { type: 'null' }],
      },
    },
    // inScope and nextQuestion are always required; the draft fields stay optional because an
    // early turn legitimately proposes nothing but a question.
    required: ['inScope', 'nextQuestion'],
    additionalProperties: false,
  }
}

function nullable(constrained: Record<string, unknown>): Record<string, unknown> {
  return { anyOf: [constrained, { type: 'null' }] }
}

/** A closed enum of the retrieved ids, or a bare null type when the organization has none - an
 * empty `enum` array is not valid JSON Schema, and "no options" must still be expressible. */
function nullableEnum(ids: readonly string[], description: string): Record<string, unknown> {
  if (ids.length === 0) {
    return { type: 'null', description }
  }

  return {
    description,
    anyOf: [{ type: 'string', enum: ids }, { type: 'null' }],
  }
}

File: src/Collega.Infrastructure/Ai/IdeaDraftSchema.cs
using System.Text.Json;
using Collega.Application.Ai;
using Collega.Domain.Enums;
using Collega.Domain.Ideas;

namespace Collega.Infrastructure.Ai;

/// <summary>
/// Builds the response JSON Schema <b>per request</b> from the retrieval result
/// (SPEC/20-feature-ai-idea-assist.md rules 15–18).
/// </summary>
/// <remarks>
/// <para><b>This is the containment mechanism.</b> Because <c>ideaTypeId</c> and <c>businessImpactId</c>
/// are <c>enum</c>s of this organization's real, active option ids, an invalid or cross-org
/// classification is <i>structurally impossible</i> — not prompt-discouraged. Rule 13 forbids
/// restating that guarantee as "the retrieved context will keep it on topic", and rule 16 forbids
/// substituting prompt instructions for it. If you are tempted to relax a constraint here, the
/// feature's security argument relaxes with it.</para>
///
/// <para><c>additionalProperties</c> is false and every enum is closed, so <c>nextQuestion</c> is the
/// only string the model can author. There is no field in which a limerick, a recipe, or a
/// general-knowledge answer could be returned.</para>
/// </remarks>
internal static class IdeaDraftSchema
{
    /// <summary>
    /// A per-request schema whose option enums are exactly the ids in <paramref name="context"/>.
    /// </summary>
    public static Dictionary<string, JsonElement> Build(IdeaAssistContext context)
    {
        var properties = new Dictionary<string, object>
        {
            ["inScope"] = new
            {
                type = "boolean",
                description =
                    "True when the user's latest message could plausibly become an idea of one of this "
                    + "organization's active idea types, within any scope statement given. False otherwise.",
            },
            ["nextQuestion"] = new
            {
                type = "string",
                description =
                    "One short follow-up question that moves the draft forward. This is the only free "
                    + "text you may write.",
            },
            ["title"] = Nullable(new { type = "string", maxLength = Idea.TitleMaxLength }),
            ["description"] = Nullable(new { type = "string", maxLength = Idea.DescriptionMaxLength }),
            ["ideaTypeId"] = NullableEnum(context.IdeaTypeIds, "The idea type, chosen from this organization's active types."),
            ["businessImpactId"] = NullableEnum(context.BusinessImpactIds, "The business impact, chosen from this organization's active options."),
            ["priority"] = new
            {
                anyOf = new object[]
                {
                    new { type = "string", @enum = Enum.GetNames<Priority>() },
                    new { type = "null" },
                },
            },
        };

        var schema = new
        {
            type = "object",
            properties,
            // inScope and nextQuestion are always required; the draft fields stay optional because an
            // early turn legitimately proposes nothing but a question (rule 19).
            required = new[] { "inScope", "nextQuestion" },
            additionalProperties = false,
        };

        return JsonSerializer
            .Deserialize<Dictionary<string, JsonElement>>(JsonSerializer.Serialize(schema))!;
    }

    /// <summary>
    /// A closed enum of the retrieved ids, or a bare null type when the organization has none — an
    /// empty <c>enum</c> array is not valid JSON Schema, and "no options" must still be expressible.
    /// </summary>
    private static object NullableEnum(IReadOnlyList<Guid> ids, string description)
    {
        if (ids.Count == 0)
        {
            return new { type = "null", description };
        }

        return new
        {
            description,
            anyOf = new object[]
            {
                new { type = "string", @enum = ids.Select(id => id.ToString()).ToArray() },
                new { type = "null" },
            },
        };
    }

    private static object Nullable(object constrained) => new
    {
        anyOf = new[] { constrained, new { type = "null" } },
    };
}
